### PR TITLE
Inside Rust post: FFI-unwind design meeting

### PR DESCRIPTION
**Pull request by BatmanAoD**
The meeting date is still TBD. I believe it will either be this coming Monday (2 March) or two weeks after that (16 March). This must be finalized before posting.

### CODE REVIEW

File: posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+the `libc` crate such as `pthread_exit` and `longjmp` will be able to unwind
+frames with destructors when compiled with `panic=unwind`, which is not possible
+with the other proposals.
+
+## Comparison table for the proposed designs
+
+In this table, "UB" stands for "undefined behavior". Some instances of UB can
+be detected at runtime, but the code to do so would impose an undesirable
+code-size penalty; the group recommends generating such code only for debug
+builds. These cases are marked "UB (debug: abort)"
```
Comment (Amanieu) [inline posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md]: We don't have any cases marked "UB (debug: abort)" in the table.

**Comment by BatmanAoD** [comment]
The table is not particularly readable; colors or something might help.

I copied the HTML generated by GitHub because it appears that the blog doesn't use GitHub's extension for generating tables delimited by `|` characters.

File: posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md (hunk 1 above)
Comment (BatmanAoD) [reply posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md]: It seems to me that if we didn't care too much about code size for debug builds, we could make all the UB entries be "debug: abort". What do you think?

Comment (BatmanAoD) [reply posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md]: ...well, no, I guess that's not true for cases where forced unwinding is allowed but destructors won't be invoked.

Comment (BatmanAoD) [reply posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md]: So would it be possible to use abort shims for all of the UB scenarios in the last column, "other foreign unwind"?

Comment (Amanieu) [reply posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md]: We can detect UB in the last column with abort shims around all FFI calls. We can detect UB from forced unwind running destructors by still emitting destructor landing pads but making them abort. In both cases we're pretty much doing codegen as if using `panic=unwind`, which might be too much of a cost even for debug builds.

Comment (BatmanAoD) [reply posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md]: In the second-to-last column, though, destructors _won't_ reliably be run, even in debug mode, right?

Comment (BatmanAoD) [reply posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md]: Actually, I guess there's no reason why they wouldn't.

Comment (BatmanAoD) [reply posts/inside-rust/2020-02-26-ffi-unwind-design-meeting.md]: Since, as you say, we'd be doing codegen as if it were `panic=unwind`

**Comment by BatmanAoD** [comment]
One other oddity: when I generate the blog site locally, it lists this entry chronologically earlier than the already-published [incident report](https://blog.rust-lang.org/inside-rust/2020/02/26/crates-io-incident-report.html). I don't know how to change this without changing the date of the draft to tomorrow.

**Comment by XAMPPRocky** [comment]
@BatmanAoD Yes, that is a known issue with the blog. I would just make the date tomorrow, it's tomorrow somewhere.

**Comment by XAMPPRocky** [comment]
@BatmanAoD Do you want to wait for more review or should I merge this?

File: posts/inside-rust/2020-02-27-ffi-unwind-design-meeting.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+supports unwinding or not. However, destructors may not be executed if the
+frames being unwound do not have unwinding support.
+
+There are two common examples of forced unwinding:
+
+* On Windows platforms, `longjmp` is implemented as a forced unwind.
+* On glibc Linux, `pthread_exit` and `pthread_cancel` are implemented as a forced unwind.
+  * In fact, `pthread_cancel` can cause all manner of C functions to unwind,
+    including common functions like `read` and `write`.  (For a complete list,
+    search for "cancellation points" in the [pthreads man page](man-pthreads).)
```
[the remark this answers is not in the document]
Comment (BatmanAoD) [reply posts/inside-rust/2020-02-27-ffi-unwind-design-meeting.md]: Oops, I forgot the syntax for non-inline links is different from the inline syntax. Thanks.

**Comment by BatmanAoD** [comment]
@XAMPPRocky Since you've written and published posts before, if it looks okay to you, then I think we're good to go. Thank you!

**Comment by nikomatsakis** [comment]
Hmm I don't see this on https://blog.rust-lang.org/inside-rust/

**Comment by XAMPPRocky** [comment]
@nikomatsakis https://blog.rust-lang.org/inside-rust/2020/02/27/ffi-unwind-design-meeting.html

**Pull request merged**